src/uUV.cpp

```diff
@@ -1,11 +1,356 @@
-#ifdef USE_MICRO_UV
-
 #include "uUV.h"
 
+#ifndef USE_MICRO_UV
+void uv_close(uv_async_t *handle, uv_close_cb cb) {
+    uv_close((uv_handle_t *) handle, cb);
+}
+void uv_close(uv_idle_t *handle, uv_close_cb cb) {
+    uv_close((uv_handle_t *) handle, cb);
+}
+void uv_close(uv_poll_t *handle, uv_close_cb cb) {
+    uv_close((uv_handle_t *) handle, cb);
+}
+void uv_close(uv_timer_t *handle, uv_close_cb cb) {
+    uv_close((uv_handle_t *) handle, cb);
+}
+#else
+
+#include <sys/eventfd.h>
+
+//namespace uUV {
+
 uv_loop_t *loops[128];
-int loopHead;
+int loopHead = 0;
+
+#define CALLBACK_ARR_SIZE 128
+uv_async_cb async_callbacks[CALLBACK_ARR_SIZE];
+int asyncCbHead = 0;
+uv_idle_cb idle_callbacks[CALLBACK_ARR_SIZE];
+int idleCbHead = 0;
+uv_poll_cb poll_callbacks[CALLBACK_ARR_SIZE];
+int pollCbHead = 0;
+uv_timer_cb timer_callbacks[CALLBACK_ARR_SIZE];
+int timerCbHead = 0;
+
+uv_loop_t *uv_handle_t::get_loop() const {
+    return loops[loopIndex];
+}
+
+inline uv_loop_t *uv_loop_helper() {
+    uv_loop_t *loop = new uv_loop_t;
+    loop->efd = epoll_create(1);
+    loop->index = loopHead++;
+    loop->numEvents = 0;
+
+    loop->asyncWakeupFd = eventfd(0, EFD_SEMAPHORE | EFD_NONBLOCK);
+    struct epoll_event wakeupEvents;
+    wakeupEvents.events = EPOLLHUP | EPOLLERR | EPOLLIN | EPOLLET;
+    wakeupEvents.data.ptr = nullptr;
+    epoll_ctl(loop->efd, EPOLL_CTL_ADD, loop->asyncWakeupFd, &wakeupEvents);
+
+    loops[loop->index] = loop;
+    return loop;
+}
+
+inline void init() {
+    uv_loop_helper();
+}
+
+uv_loop_t *uv_default_loop() {
+    if (!loopHead) {
+        init();
+    }
+    return loops[0];
+}
+
+uv_loop_t *uv_loop_new() {
+    if (!loopHead)
+        init();
+    return uv_loop_helper();
+}
+
+void uv_loop_delete(uv_loop_t *loop) {
+    epoll_ctl(loop->efd, EPOLL_CTL_DEL, loop->asyncWakeupFd, nullptr);
+    close(loop->efd);
+    loops[loop->index] = nullptr;
+    delete loop;
+}
+
+bool uv_is_closing(uv_handle_t *handle) {
+    return handle->flags & (UV_HANDLE_CLOSING | UV_HANDLE_CLOSED);
+}
+
+void uv_async_init(uv_loop_t *loop, uv_async_t *async, uv_async_cb cb) {
+    async->loopIndex = loop->index;
+    loop->numEvents++;
+
+    async->cbIndex = asyncCbHead;
+    for (int i = 0; i < asyncCbHead; i++) {
+        if (async_callbacks[i] == cb) {
+            async->cbIndex = i;
+            break;
+        }
+    }
+    if (async->cbIndex == asyncCbHead) {
+        async_callbacks[asyncCbHead++] = cb;
+    }
+
+    loop->asyncs.insert(async);
+}
+
+void uv_async_send(uv_async_t *async) {
+    uv_loop_t *loop = async->get_loop();
+    loop->async_mutex.lock();
+    uint64_t val = 1;
+    int w = write(loop->asyncWakeupFd, &val, sizeof(val));
+    async->run = true;
```

Comment:
run is needed to identify which async handles are ready to fire. The mutex and the eventfd wakeup event are shared for all of the asyncs in a loop, so there needs to be something to tell us which subset of asyncs to call.